### PR TITLE
net: Add file_socket function

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
and move the socket special process from fstat/nx_vfcntl/ to file_fstat/file_vfcntl.
This PR also simplify sendfile implementation.

## Impact
Make the behaviour is smae between fstat/fcntl and file_fstat/file_fcntl

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@gustavonihei could you review this PR?

File: net/socket/net_sendfile.c [hunk 1]
```diff
@@ -130,8 +130,8 @@ ssize_t psock_sendfile(FAR struct socket *psock, FAR struct file *infile,
   if (psock == NULL || psock->s_conn == NULL)
     {
       nerr("ERROR: Invalid socket\n");
-      _SO_SETERRNO(psock, EBADF);
-      return ERROR;
+      psock->s_error = EBADF;
```
Comment (gustavonihei) [inline net/socket/net_sendfile.c:133]: `psock_sendfile` documentation must be updated to reflect the current implementation. The __Return Value__ section still states that on failure it returns -1 and sets `errno`.

File: fs/vfs/fs_sendfile.c [hunk 2]
```diff
@@ -26,20 +26,255 @@
...
+
+  /* Release the I/O buffer */
+
+  kmm_free(iobuffer);
+
+  /* Return the current file position */
+
+  if (offset)
+    {
+      /* Use lseek to get the current file position */
```
Comment (gustavonihei) [inline fs/vfs/fs_sendfile.c:203]:
```suggestion
      /* Use file_seek to get the current file position */
```

File: fs/vfs/fs_sendfile.c [hunk 3]
```diff
@@ -26,20 +26,255 @@
...
+      if (curpos < 0)
+        {
+          return curpos;
+        }
+
+      /* Return the current file position */
+
+      *offset = curpos;
+
+      /* Use lseek again to restore the original file position */
```
Comment (gustavonihei) [inline fs/vfs/fs_sendfile.c:215]:
```suggestion
      /* Use file_seek again to restore the original file position */
```

File: fs/vfs/fs_sendfile.c [hunk 4]
```diff
@@ -26,20 +26,255 @@
...
+           */
+
+          wrbuffer = iobuffer;
+          do
+            {
+              /* Write the buffer of data to the outfile */
+
+              nbyteswritten = file_write(outfile, wrbuffer, nbytesread);
+
+              /* Check for a complete (or parial) write.  write() should not
```
Comment (gustavonihei) [inline fs/vfs/fs_sendfile.c:147]:
```suggestion
              /* Check for a complete (or partial) write.  write() should not
```

File: fs/vfs/fs_sendfile.c [hunk 5]
```diff
@@ -26,20 +26,255 @@
...
+  size_t  ntransferred;
+  bool endxfr;
+
+  /* Get the current file position. */
+
+  if (offset)
+    {
+      off_t newpos;
+
+      /* Use lseek to get the current file position */
```
Comment (gustavonihei) [inline fs/vfs/fs_sendfile.c:55]:
```suggestion
      /* Use file_seek to get the current file position */
```

File: fs/vfs/fs_sendfile.c [hunk 6]
```diff
@@ -26,20 +26,255 @@
...
+
+      startpos = file_seek(infile, 0, SEEK_CUR);
+      if (startpos < 0)
+        {
+          return startpos;
+        }
+
+      /* Use lseek again to set the new file position */
```
Comment (gustavonihei) [inline fs/vfs/fs_sendfile.c:63]:
```suggestion
      /* Use file_seek again to set the new file position */
```

**Comment by gustavonihei** [comment]
I think overall it looks good.
I've seen some other PRs with the same approach of changing the file operations to use `filep` structure instead of file descriptors. Could you explain the motivation for this refactor?
My first impression it seems that it looks cleaner, but I'd like to have a better comprehension of this.

File: net/socket/net_sendfile.c (hunk 1 above)
Comment (xiaoxiang781216) [reply net/socket/net_sendfile.c:133]: Done.

File: fs/vfs/fs_sendfile.c (hunk 2 above)
Comment (xiaoxiang781216) [reply fs/vfs/fs_sendfile.c]: Done.

File: fs/vfs/fs_sendfile.c (hunk 3 above)
Comment (xiaoxiang781216) [reply fs/vfs/fs_sendfile.c]: Done.

File: fs/vfs/fs_sendfile.c (hunk 4 above)
Comment (xiaoxiang781216) [reply fs/vfs/fs_sendfile.c]: Done.

File: fs/vfs/fs_sendfile.c (hunk 5 above)
Comment (xiaoxiang781216) [reply fs/vfs/fs_sendfile.c]: Done.

File: fs/vfs/fs_sendfile.c (hunk 6 above)
Comment (xiaoxiang781216) [reply fs/vfs/fs_sendfile.c]: Done.

**Comment by xiaoxiang781216** [comment]
File descriptor always associate with a particular task, but kernel thread or function should be task independent as much as possible. This patch series try to limit file descriptor function as a thin wrapper around file structure function.

**Comment by gustavonihei** [comment]
> File descriptor always associate with a particular task, but kernel thread or function should be task independent as much as possible. This patch series try to limit file descriptor function as a thin wrapper around file structure function.

True, it makes sense.

File: net/socket/net_sendfile.c [hunk 7]
```diff
@@ -74,7 +74,7 @@
  *
  * Returned Value:
  *   On success, returns the number of characters sent.  On  error,
- *   -1 is returned, and errno is set appropriately:
+ *   the negative errno is return appropriately:
```
Comment (gustavonihei) [inline net/socket/net_sendfile.c:77]:
```suggestion
 *   the negative errno is returned appropriately:
```
nit: Typo

Comment (xiaoxiang781216) [reply net/socket/net_sendfile.c:77]: Done, thanks.

**Review by gustavonihei — APPROVED** [review]

**Pull request merged**